### DISCUSSION TITLE
Adjusting image ratio trial by trial

### DISCUSSION THREAD

**Discussion by marco2gandolfo** [jspsych/jsPsych#1204] [Q&A] [answered]
Hello all,

I have an image set with images with different size ratios. In the experiments each of this images will be followed by the presentation of a mask. What I would like to achieve is to adjust the image ratio of the mask depending on the size of the previously presented image. 
Is there a way to extract the image size using the onfinish or on_start event of the image-keyboard_response trial related to the stimulus to be used in the next image_keyboard response related to the mask? As you would do for example for feedback for trials on accuracy?

Many thanks for your help.

Marco

**Comment by jodeleeuw** [comment, marked as answer]
Hi Marco,

In the `on_load` parameter for a trial you can access information about the objections currently on the screen. If you have one image on the screen then something like this should work:

```js
on_load: function(){
	var image = document.querySelector('img');
	var width = image.clientWidth;
	var height = image.clientHeight;
}
```

The tricky part is getting this information into the trial data, since the data isn't easily accessible from within the `on_load` event.

You could create a global variable for `height` and `width` and set this in `on_load` and then use it in `on_finish` to modify the data.

```js
var height;
var width;
var trial = {
	on_load: function(){
		var image = document.querySelector('img');
		width = image.clientWidth;
		height = image.clientHeight;
	},
	on_finish: function(data){
		data.imgHeight = height;
		data.imgWidth = width;
	}
}
```

**Reply by marco2gandolfo** [reply]
Dear @jodeleeuw ,

Thanks a lot for your prompt reply!
This seems the perfect solution! However, at the moment it is returning 0. I have also tried image.naturalWidth and image.naturalHeight but it does the same.

It seems to not be able to retrieve the img correctly. So I thought that maybe it matters the structure of my script? Particularly I use as below a trial sequence that calls timeline variables, maybe this makes harder for onload to access info on the image?

```javascript
var my_trial_seq = {
                    type: 'image-keyboard-response',
                    on_load: function(){
                        var image = document.querySelector('img');
		                    width = image.clientWidth;
		                    height = image.clientHeight;
                    },
                    stimulus: jsPsych.timelineVariable('stimulus'),
                    choices: jsPsych.NO_KEYS,
                    on_start: function () { $('body').css('cursor', 'none') }, // hide cursor where not needed
                    on_finish: function(data){ 
                        data.imgHeight = height;
		        data.imgWidth = width;
                    console.log("image width is", data.imgWidth, "image height is", data.imgHeight)
                    },
                    trial_duration: 250
                },
                {
                    type: 'image-keyboard-response',
                    stimulus: jsPsych.randomization.sampleWithReplacement(mask_array, 1),
                    choices: jsPsych.NO_KEYS,
                    on_start: function () { $('body').css('cursor', 'none') }, // hide cursor where not needed
                    trial_duration: 50
                }
```

**Reply by becky-gilbert** [reply]
Hi @marco2gandolfo, are you manually preloading the image files? If not then I think that's the problem. I think what's happening is that, when the trial's `on_load` function is called, the image data still needs to be transferred, so the clientWidth and clientHeight information isn't available yet. (I think the `on_load` function is called when the page is updated with the trial's HTML content, but not necessarily when everything on the page has finished loading).
Also, in case you weren't aware: media files are often automatically preloaded when the file name is passed directly to the plugin stimulus parameter, but if you're using timeline variables then they have to be [manually preloaded](https://www.jspsych.org/overview/media-preloading/#manual-preloading).
Here's an example that's working for me:
```javascript
    var height, width;
    var stim_info = [
        {stimulus: 'jspsych-6.1.0/examples/img/1.gif'},
        {stimulus: 'jspsych-6.1.0/examples/img/backwardN.gif'}
    ];
    var mask_array = ['jspsych-6.1.0/examples/img/2.gif','jspsych-6.1.0/examples/img/3.gif'];
    var my_trial_seq = [
        {
            type: 'image-keyboard-response',
            on_load: function(){
                var image = document.querySelector('img');
                width = image.clientWidth;
                height = image.clientHeight;
            },
            stimulus: jsPsych.timelineVariable('stimulus'),
            choices: jsPsych.NO_KEYS,
            //on_start: function () { $('body').css('cursor', 'none') }, // hide cursor where not needed
            on_finish: function(data){ 
                data.imgHeight = height;
                data.imgWidth = width;
                console.log("image width is", data.imgWidth, "image height is", data.imgHeight)
            },
            trial_duration: 250
        },
        {
            type: 'image-keyboard-response',
            stimulus: jsPsych.randomization.sampleWithReplacement(mask_array, 1)[0],
            choices: jsPsych.NO_KEYS,
            //on_start: function () { $('body').css('cursor', 'none') }, // hide cursor where not needed
            trial_duration: 50
        }
    ];
    var task = {
        timeline: my_trial_seq,
        timeline_variables: stim_info
    };

    jsPsych.init({
        timeline: [task],
        preload_images: ['jspsych-6.1.0/examples/img/1.gif','jspsych-6.1.0/examples/img/backwardN.gif','jspsych-6.1.0/examples/img/2.gif','jspsych-6.1.0/examples/img/3.gif'],
        on_finish: function () {
            jsPsych.data.displayData();
        }
    });
```

**Reply by marco2gandolfo** [reply]
Dear @becky-gilbert , 

Thanks a lot for this! indeed that was it! I should have thought about it. I do preload the images usually but while debugging I left it // commented // to avoid long loading times.

Great!

Thanks a lot to you and Josh for the help,

Cheers,
Marco
